### PR TITLE
Add CUAV smart battery drive

### PR DESCRIPTION
**Pull request by mxiaogit**
The modifications are as follows:
  - Add time remaining(MAVLink2 BATTERY_STATUS message support)
  - Add CUAV smart battery drive

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 1]
```diff
@@ -386,6 +392,17 @@ uint8_t AP_BattMonitor::capacity_remaining_pct(uint8_t instance) const
     }
 }
 
+/// time_remaining - returns remaining battery time
+bool AP_BattMonitor::time_remaining(int32_t &min, uint8_t instance) const
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:396]: This should be seconds.

And presumably unsigned?

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 2]
```diff
@@ -386,6 +392,17 @@ uint8_t AP_BattMonitor::capacity_remaining_pct(uint8_t instance) const
...
+{
+    if (instance < _num_instances && drivers[instance] != nullptr && drivers[instance]->has_time_remaining()) {
+        min = state[instance].time_remaining;
+        return true;
+    } else {
+        return false;
+    }
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:403]:
```suggestion
    }
    return false;
```

**Review by peterbarker — COMMENTED** [review]
I've only briefly looked at this, but it looks pretty good at first glance.

I do have to query whether we need another backend, however.  What was your reasoning for not just subscribing to the additional/alternate message in the existing `AP_BattMonitor_UAVCAN` backend?

**Comment by mxiaogit** [comment]
> I've only briefly looked at this, but it looks pretty good at first glance.
> 
> I do have to query whether we need another backend, however. What was your reasoning for not just subscribing to the additional/alternate message in the existing `AP_BattMonitor_UAVCAN` backend?
@peterbarker Where are you referring to? I'm a little confused.

**Comment by gitfishup** [comment]
> I've only briefly looked at this, but it looks pretty good at first glance.
> 
> I do have to query whether we need another backend, however. What was your reasoning for not just subscribing to the additional/alternate message in the existing `AP_BattMonitor_UAVCAN` backend?

The "another backend" means the following virtual function?
` virtual bool has_time_remaining() const { return false; }`

**Comment by gitfishup** [comment]
> > I've only briefly looked at this, but it looks pretty good at first glance.
> > I do have to query whether we need another backend, however. What was your reasoning for not just subscribing to the additional/alternate message in the existing `AP_BattMonitor_UAVCAN` backend?
> 
@peterbarker  The "another backend" means the following virtual function?
 ` virtual bool has_time_remaining() const { return false; }`

**Comment by amilcarlucas** [comment]
I did https://github.com/ArduPilot/ardupilot/pull/15410 a while ago, that one passes CI. AFAIK you need to use my code in order to get CI to pass. The log message descritipon string that you are using is too long.

**Comment by mxiaogit** [comment]
> I did #15410 a while ago, that one passes CI. AFAIK you need to use my code in order to get CI to pass. The log message descritipon string that you are using is too long.

@amilcarlucas Thank you for your help.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 1 above)
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp:396]: The type of time_remaining in the MAVLink2 BATTERY_STATUS message is int32_t. Would it be better to use this type?

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp [hunk 3]
```diff
@@ -114,15 +130,50 @@ void AP_BattMonitor_UAVCAN::handle_battery_info(const BattInfoCb &cb)
...
+
+    if (!isnanf(cb.msg->temperature) && cb.msg->temperature > 0) {
+        // Temperature reported from battery in kelvin and stored internally in Celsius.
+        _interim_state.temperature = cb.msg->temperature - C_TO_KELVIN;
+        _interim_state.temperature_time = AP_HAL::millis();
+    }
+
+    for (uint8_t i = 0; i < cb.msg->cell_count; i++) {
+        _interim_state.cell_voltages.cells[i] = cb.msg->voltage_cell[i] * 1000;
+    }
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:151]: This is a memory over-write waiting to happen; you need to constrain your loop to `ARRAY_SIZE(_interim_state.cell_voltages.cells[i])` as well as `cb.msg->cell_count`

**Review by peterbarker — CHANGES_REQUESTED** [review]
All new code must be protected with the define, as described.

@bugobliterator @magicrub @WickedShell should we really have a battery backend per uavcan message type received?  Having separate backend types means we don't need to allocate a listener for every packet type in every uavcan battery backend, but it does mean the user has to know which specific UAVCAN message they can expect from the device which kind of sucks.

I suggest we could listen for every message type for a while and then choose the best one and kill the listeners for the other.  In the short term we'd just listen for all of them and kill the listeners for ones we haven't received as soon as we get one....

**Comment by magicrub** [comment]
Generally speaking, no. We should not have multiple UAVCAN drivers. Too much redundant code. I'll take a look at this before the dev call.

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp [hunk 4]
```diff
@@ -41,18 +50,25 @@ void AP_BattMonitor_UAVCAN::subscribe_msgs(AP_UAVCAN* ap_uavcan)
     const int battinfo_listener_res = battinfo_listener->start(BattInfoCb(ap_uavcan, &handle_battery_info_trampoline));
     if (battinfo_listener_res < 0) {
         AP_HAL::panic("UAVCAN BatteryInfo subscriber start problem\n\r");
-        return;
+    }
+
+    uavcan::Subscriber<cuav::equipment::power::CBAT, CBattCb> *cbat_listener;
```
[the remark this answers is not in the document]
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:55]: Is there any way to subscribe to messages based on type?

[the remark this answers is not in the document]
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:55]: Is there any way to subscribe to messages based on type?

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 5]
```diff
@@ -87,6 +87,7 @@ class AP_BattMonitor
         GENERATOR_FUEL             = 18,
         Rotoye                     = 19,
         MPPT_PacketDigital         = 20,
+        UAVCAN_CBAT                = 21,
```
[the remark this answers is not in the document]
Comment (bugobliterator) [reply libraries/AP_BattMonitor/AP_BattMonitor.h:95]: I think better place would be `AP_HAL/AP_HAL_Boards.h`

**Comment by tridge** [comment]
what rate do you plan to send this at? It is a very big msg. Could you split it into high rate and low rate msgs?
could the individual cell voltages be float16?

**Comment by mxiaogit** [comment]
> what rate do you plan to send this at? It is a very big msg. Could you split it into high rate and low rate msgs?
> could the individual cell voltages be float16?

@tridge This message is published at a rate of about 1~2Hz.I cannot split this message into two messages.I will modify the data type of cell voltages to float16.Are there any suggestions for the data types of other fields in this message?

File: libraries/AP_BattMonitor/AP_BattMonitor.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor.h:91]: Thank you for your suggestion, I have modified it.

**Pull request closed without merging**